drape_frontend/text_layout.cpp

```diff
@@ -291,69 +300,75 @@ double GetTextMinPeriod(double pixelTextLength)
 {
-  ASSERT(!m_metrics.empty(), ());
+  ASSERT(!m_glyphRegions.empty(), ());
 #ifdef DEBUG
-  ref_ptr<dp::Texture> tex = m_metrics[0].GetTexture();
-  for (GlyphRegion const & g : m_metrics)
-  {
+  ref_ptr<dp::Texture> tex = m_glyphRegions[0].GetTexture();
+  for (GlyphRegion const & g : m_glyphRegions)
     ASSERT(g.GetTexture() == tex, ());
-  }
 #endif
 
-  return m_metrics[0].GetTexture();
+  return m_glyphRegions[0].GetTexture();
 }
 
-uint32_t TextLayout::GetGlyphCount() const
+size_t TextLayout::GetGlyphCount() const
 {
-  return static_cast<uint32_t>(m_metrics.size());
+  ASSERT_EQUAL(m_shapedGlyphs.m_glyphs.size(), m_glyphRegions.size(), ());
+  return m_glyphRegions.size();
 }
 
 float TextLayout::GetPixelLength() const
 {
-  return m_textSizeRatio * std::accumulate(m_metrics.begin(), m_metrics.end(), 0.0f,
-                                           [](double const & v, GlyphRegion const & glyph) -> float
-  {
-    return static_cast<float>(v) + glyph.GetAdvanceX();
-  });
+  // TODO(AB): Is ratio needed here?
+  return m_shapedGlyphs.m_lineWidthInPixels * m_textSizeRatio;
+  // return m_textSizeRatio * std::accumulate(m_glyphRegions.begin(), m_glyphRegions.end(), 0.0f,
+  //                                          [](double const & v, GlyphRegion const & glyph) -> float
+  // {
+  //   return static_cast<float>(v) + glyph.GetAdvanceX();
+  // });
 }
 
 float TextLayout::GetPixelHeight() const
 {
+  //return m_shapedGlyphs.m_maxLineHeightInPixels * m_textSizeRatio;
   return m_textSizeRatio * dp::kBaseFontSizePixels;
 }
 
-strings::UniString const & TextLayout::GetText() const
+dp::TGlyphs TextLayout::GetGlyphs() const
 {
-  return m_text;
+  // TODO(AB): Can conversion to TGlyphs be avoided?
+  dp::TGlyphs glyphs;
+  glyphs.reserve(m_shapedGlyphs.m_glyphs.size());
+  for (auto const & glyph : m_shapedGlyphs.m_glyphs)
+    glyphs.emplace_back(glyph.m_fontIndex, glyph.m_glyphId);
+  return glyphs;
 }
 
-StraightTextLayout::StraightTextLayout(strings::UniString const & text, float fontSize,
-                                       ref_ptr<dp::TextureManager> textures, dp::Anchor anchor, bool forceNoWrap)
+StraightTextLayout::StraightTextLayout(std::string const & text, float fontSize, ref_ptr<dp::TextureManager> textures,
+                                       dp::Anchor anchor, bool forceNoWrap)
 {
-  strings::UniString visibleText = bidi::log2vis(text);
-  // Possible if name has strange symbols only.
-  if (visibleText.empty())
-    return;
-
+  // TODO(AB): Use ICU's BreakIterator to split text properly in different languages.
+  // TODO(AB): Here the text is optionally split by its length into two pieces.
   buffer_vector<size_t, 2> delimIndexes;
-  if (visibleText == text && !forceNoWrap)
-    SplitText(visibleText, delimIndexes);
-  else
-    delimIndexes.push_back(visibleText.size());
+  // if (visibleText == text && !forceNoWrap)
+  //   SplitText(visibleText, delimIndexes);
+  // else
```

Comment:
What is the problem with keeping an existing SplitText logic? How does it break HB?

Reply:
It does not break, just a bit more complex computation. I will implement it.

@vng can you please review newly added commits?